overlord/managers_test.go

```diff
@@ -8693,3 +8694,112 @@ volumes:
 	//  gadget content is not updated because there is no edition update
 	c.Check(filepath.Join(dirs.GlobalRootDir, "/run/mnt/", structureName, "start.elf"), testutil.FileContains, "start.elf rev1")
 }
+
+// deal with the missing "update-gadget-assets" tasks, see LP:#1940553
+func (ms *gadgetUpdatesSuite) TestGadgetKernelRefreshFromOldBrokenSnap(c *C) {
+	st := ms.o.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// we have an install kernel/gadget
+	gadgetSnapYaml := "name: pi\nversion: 1.0\ntype: gadget"
+	ms.mockInstalledSnapWithFiles(c, gadgetSnapYaml, [][]string{
+		{"meta/gadget.yaml", "volumes:\n volume-id:\n  bootloader: grub"},
+	})
+	kernelSnapYaml := "name: pi-kernel\nversion: 1.0\ntype: kernel"
+	ms.mockInstalledSnapWithFiles(c, kernelSnapYaml, nil)
+
+	// add new kernel/gadget snap
+	newKernelSnapYaml := kernelSnapYaml
+	ms.mockSnapUpgradeWithFiles(c, newKernelSnapYaml, nil)
+	ms.mockSnapUpgradeWithFiles(c, gadgetSnapYaml, [][]string{
+		{"meta/gadget.yaml", "volumes:\n volume-id:\n  bootloader: grub"},
+	})
+
+	// now a refresh is simulated that does *not* contain an
+	// "update-gadget-assets" task, see LP:#1940553
+	affected, tasksets, err := snapstate.UpdateMany(context.TODO(), st, nil, 0, &snapstate.Flags{})
+	c.Assert(err, IsNil)
+	sort.Strings(affected)
+	c.Check(affected, DeepEquals, []string{"pi", "pi-kernel"})
+
+	// here we need to manipulate the change to simulate that there
+	// is no "update-gadget-assets" task for the kernel, unfortunately
+	// there is no "state.TaskSet.RemoveTask" nor a "state.Task.Unwait()"
+	chg := st.NewChange("upgrade-snaps", "...")
+	for _, ts := range tasksets {
+		// skip the taskset of UpdateMany that does the
+		// check-rerefresh, see tsWithoutReRefresh for details
+		if ts.Tasks()[0].Kind() == "check-rerefresh" {
+			continue
+		}
+
+		// Remove the "update-gadget-assets" task from the kernel
+		// update TaskSet to simulate an upgrade from an old snapd
+		// (see bug https://bugs.launchpad.net/snapd/+bug/1940553)
+		snapsup, err := snapstate.TaskSnapSetup(ts.Tasks()[0])
+		c.Assert(err, IsNil)
+		if snapsup.Type == "kernel" {
```

Comment:
it's clear you're already aware of this, but this code is a bit fragile if we change the set of tasks for the install to be more complicated :-/

then again given the unending problems we have with tasks and old snapds doing the wrong thing and such, perhaps we will never change this precisely because of this sort of problem :man_shrugging:

Reply:
Nice idea! Thanks! But passing this down is also surprisingly fiddly in the main code but maybe I am approaching it wrong, I will continue after my next meeting.

Reply:
sorry. my wording isn't so clear, what I meant is bluntly: var TestingLeaveOutKernetUpdateGadgetAssets bool = false

Reply:
Oh, very sorry, I misunderstood, fixingthis now

Reply:
This is done now.